types/dec_coin.go

```diff
@@ -157,9 +157,20 @@ func (coin DecCoin) IsValid() bool {
 // DecCoins defines a slice of coins with decimal values
 type DecCoins []DecCoin
 
-// NewDecCoins constructs a new coin set with decimal values
+// NewDecCoins constructs a new coin set with with decimal values
+// from DecCoins.
+func NewDecCoins(decCoinArray ...DecCoin) DecCoins {
```

Comment:
We need to add extra checks for the coins that are being created with the constructor:

- remove empty (zero value) DecCoins
- sort coins by denomination
- panic on duplicated denomination
- panic on invalid decCoin

You can check the `NewCoins` constructor for reference

Reply:
I tried to make the findDup a bit more generic to use for Coins and DecCoins with the same function, but I'm not a big fan of interfaces for these solutions of reference types because of the heap allocations of them. But if you agree with that it's good to me.